include/pybind11/numpy.h

```diff
@@ -316,7 +316,64 @@ class dtype : public object {
     }
 };
 
-class array : public buffer {
+namespace detail {
+    inline void fail_dim_check(size_t dim, size_t ndim, const std::string& msg) {
+        throw index_error(msg + ": " + std::to_string(dim) +
+                          " (ndim = " + std::to_string(ndim) + ")");
+    }
+}
+
+class safe_access_policy {
+public:
+    void check_axis(size_t dim, size_t ndim) const {
+        if(dim >= ndim) {
+            detail::fail_dim_check(dim, ndim, "invalid axis");
+        }
+    }
+
+    template <typename... Ix>
+    void check_indices(size_t ndim, Ix...) const
+    {
+        if(sizeof...(Ix) > ndim) {
+            detail::fail_dim_check(sizeof...(Ix), ndim, "too many indices for an array");
+        }
+    }
+
+    template<typename... Ix>
+    void check_dimensions(const size_t* shape, Ix... index) const {
+        check_dimensions_impl(size_t(0), shape, size_t(index)...);
+    }
+
+private:
+    void check_dimensions_impl(size_t, const size_t*) const { }
+
+    template<typename... Ix>
+    void check_dimensions_impl(size_t axis, const size_t* shape, size_t i, Ix... index) const {
+        if (i >= *shape) {
+            throw index_error(std::string("index ") + std::to_string(i) +
+                              " is out of bounds for axis " + std::to_string(axis) +
+                              " with size " + std::to_string(*shape));
+        }
+        check_dimensions_impl(axis + 1, shape + 1, index...);
+    }
+};
+
+class unsafe_access_policy {
+public:
+    void check_axis(size_t, size_t) const {
+    }
+
+    template <typename... Ix>
+    void check_indices(size_t, Ix...) const {
+    }
+
+    template <typename... Ix>
+    void check_dimensions(const size_t*, Ix...) const {
+    }
+};
+
+template <class access_policy = safe_access_policy>
+class array : public buffer, private access_policy {
```

Comment:
Changing `array` to a template class is definitely going to break code; this needs to be somewhere new with a `using array = array_base<safe_access_policy>`.  With that, you might as well drop the default on the template argument as well.  You might also add a `using array_unchecked = array_base<unsafe_access_policy>;`